ProjFS.Mac/PrjFSKext/Memory.hpp

```diff
@@ -18,21 +18,25 @@ void Memory_Free(void* buffer, uint32_t size);
 template <typename T>
 T* Memory_AllocArray(uint32_t arrayLength)
 {
-    size_t allocBytes = arrayLength * sizeof(T);
-    if (allocBytes > UINT32_MAX)
+    uint32_t allocBytes;
+    if (__builtin_umul_overflow(arrayLength, sizeof(T), &allocBytes))
     {
+        // Overflow occurred.
         return nullptr;
     }
-    
-    return static_cast<T*>(Memory_Alloc(static_cast<uint32_t>(allocBytes)));
+
+    return static_cast<T*>(Memory_Alloc(allocBytes));
 }
 
 template <typename T>
 void Memory_FreeArray(T* array, uint32_t arrayLength)
```

Comment:
Unit Test?

Reply:
Done, though this required changing the signature. The other alternative would be to provide a mocked implementation of `assert()`. @pmj @jeschu1 any preferences?

Reply:
Yes, I agree, let's not change the return type. The in-kext asserts are currently badly integrated with the unit tests, but this is something I'd like to fix.

The size calculations in `Memory_AllocArray` and `Memory_FreeArray` are identical, so perhaps we can pull those out into a helper function and unit test that. You could also leave a TODO comment for testing `Memory_FreeArray` once we can handle in-kext asserts in unit tests.